cmd/launcher/doctor.go

```diff
@@ -0,0 +1,531 @@
+	"github.com/kolide/launcher/pkg/autoupdate"
+	"github.com/kolide/launcher/pkg/autoupdate/tuf"
+	"github.com/kolide/launcher/pkg/launcher"
+	"github.com/kolide/launcher/pkg/log/checkpoint"
+	"github.com/peterbourgon/ff/v3"
+
+	"github.com/fatih/color"
+	"golang.org/x/exp/slices"
+)
+
+var (
+	cRunning = color.New(color.FgCyan)
+	cHeader  = color.New(color.FgHiWhite).Add(color.Bold)
+	cWarn    = color.New(color.FgHiYellow)
+
+	pass = color.New(color.FgGreen).PrintlnFunc()
+	fail = color.New(color.FgRed).Add(color.Bold).PrintlnFunc()
+
+	// Println functions for checkup details
+	cInfo = color.New(color.FgWhite)
+	info  = func(a ...interface{}) {
+		cInfo.Println(fmt.Sprintf("    %s", a...))
+	}
+	warn = func(a ...interface{}) {
+		cWarn.Println(fmt.Sprintf("    %s", a...))
+	}
+	bad = func(a ...interface{}) {
+		cInfo.Println(fmt.Sprintf("❌  %s", a...))
+	}
+	good = func(a ...interface{}) {
+		cInfo.Println(fmt.Sprintf("✅  %s", a...))
+	}
+
+	configFile string
+)
+
+// checkup is
+type checkup struct {
+	name   string
+	check  func() (string, error)
+	failed bool
+}
+
+func runDoctor(args []string) error {
+	logger := log.With(logutil.NewCLILogger(true), "caller", log.DefaultCaller)
+	opts, err := parseDoctorOptions(os.Args[2:])
+	if err != nil {
+		level.Info(logger).Log("err", err)
+		os.Exit(1)
+	}
+
+	fcOpts := []flags.Option{flags.WithCmdLineOpts(opts)}
+	flagController := flags.NewFlagController(logger, nil, fcOpts...)
+	k := knapsack.New(nil, flagController, nil)
+
+	cRunning.Println("Kolide launcher doctor version:")
+	version.PrintFull()
+	cRunning.Println("\nRunning Kolide launcher checkups...")
+
+	checkups := []*checkup{
+		{
+			name: "Check platform",
+			check: func() (string, error) {
+				return checkupPlatform(runtime.GOOS)
+			},
+		},
+		{
+			name: "Root directory contents",
+			check: func() (string, error) {
+				return checkupRootDir(getFilepaths(k.RootDirectory(), "*"))
+			},
+		},
+		// TODO: Check application directories for launcher & osquery binaries
+		{
+			name: "Check communication with Kolide",
+			check: func() (string, error) {
+				return checkupConnectivity(logger, k)
+			},
+		},
+		{
+			name: "Check version",
+			check: func() (string, error) {
+				return checkupVersion(version.Version())
+			},
+		},
+		{
+			name: "Check config file",
+			check: func() (string, error) {
+				return checkupConfigFile(configFile)
+			},
+		},
+		{
+			name: "Check logs",
+			check: func() (string, error) {
+				return checkupLogFiles(getFilepaths(k.RootDirectory(), "debug*"))
+			},
+		},
+	}
+
+	failedCheckups := []*checkup{}
+
+	// Sequentially run all of the checkups
+	for _, c := range checkups {
+		c.run()
+	}
+
+	if len(failedCheckups) > 0 {
+		fail("\nSome checkups failed:")
+
+		for _, c := range failedCheckups {
+			fail("%s\n", c.name)
+		}
+	} else {
+		pass("\nAll checkups passed! Your Kolide launcher is healthy.")
+	}
+
+	return nil
+}
+
+func parseDoctorOptions(args []string) (*launcher.Options, error) {
```

Comment:
I think some consolidation makes sense -- I would be worried about getting out of sync between parseOptions and this one when adding a new command-line flag

Reply:
Sadly, I'm not sure we have great options.

I think we're carrying a lot of weird baggage because:
* launcher have a lot of options
* we use go's flags library
* flags does not support ignoring unknown options

So I guess we maybe only have a couple choices. We could push flag parsing into something shared.

Or we could use a different flag/config file parser for things like `doctor` which have pretty narrow needs.

Reply:
Though some of this could, maybe, be absorbed by knapsack's flag stuff. Not sure?